### ISSUE TITLE
[Feature Request] Automatic mixed precision training.

### ISSUE THREAD

**Issue by lkhphuc** [poets-ai/elegy#224]
**Is your feature request related to a problem? Please describe.**
Mixed precision training is getting more and more necessary and is being make more and more accessible in every framework. 
In jax, Deepmind's [JMP](https://github.com/deepmind/jmp) is definitely the way to go.
Make it work seamlessly with Elegy would be great. 

**Describe the solution you'd like**
#### As a flag in `model.fit`
Given there's already a `eager` flag, elegy could support different JMP's policy by implementing a default policy of full precision. User will define the policy and pass that in as an argument to the `fit` method similar to `eager` flag.
Disadvantages: 
- Might add a small overhead by adding a default full precision policy
- Might not be able to support Loss scaling.

#### As a Callback
Just provide a default implementation as a callback.

#### Just use low-level API
- Advantage: No modification, just provide an example.
- Disadvantage: Must modify the `train_step` method to use loss scaling.

Curious to hear your opinion. I'm trying the low-level API first and will update with an example soon.

**Comment by lkhphuc** [comment]
I modified an example with low-level API here.
```
import elegy as eg
...

M = tp.TypeVar("M", bound="Model")
half = jnp.float16
full = jnp.float32

class Model(eg.Model):
    module: eg.Module = eg.Parameter.node()
    policy: jmp.Policy

    def __init__(
        self,
        features_out: int,
        loss: tp.Any = None,
        metrics: tp.Any = None,
        optimizer=None,
        seed: int = 42,
        eager: bool = False,
    ):
        self.features_out = features_out
        super().__init__(
            loss=loss,
            metrics=metrics,
            optimizer=optimizer,
            seed=seed,
            eager=eager,
        )
        self.policy: jmp.Policy = jmp.Policy(compute_dtype=half, param_dtype=full, output_dtype=half)
        # self.policy: jmp.Policy = jmp.Policy( compute_dtype=full, param_dtype=full, output_dtype=full)
        self.loss_scale: jmp.LossScale = jmp.StaticLossScale(2 ** 2)

    def init_step(self: M, key: jnp.ndarray, inputs: jnp.ndarray) -> M:
        model: M = self
        features_in = np.prod(inputs.shape[1:])

        inputs = E.rearrange(inputs, "b h w -> b (h w)")
        self.module = eg.MLP(
            [ features_in,  2 ** 12,  2 ** 8,  2 ** 15,  2 ** 11, 2 ** 13, self.features_out ]
        ).init(key, inputs)

        assert self.optimizer is not None
        self.optimizer = self.optimizer.init(self)

        params = self.policy.cast_to_param(model.parameters())
        model = model.merge(params)

        return model

    def pred_step(self: M, inputs: tp.Any) -> eg.PredStepOutput[M]:
        model = self
        inputs, model = self.policy.cast_to_compute((inputs, model))
        logits = model(inputs)
        return logits, model

    def test_step(
        self,
        inputs,
        labels,
    ):
        model = self
        # flatten + scale
        inputs = jnp.reshape(inputs, (inputs.shape[0], -1)) / 255

        # forward
        logits, model = model.pred_step(inputs)

        # crossentropy loss
        target = jax.nn.one_hot(labels["target"], self.features_out)
        target = self.policy.cast_to_output(target)

        loss = optax.softmax_cross_entropy(logits, target).mean()

        # metrics
        logs = dict(
            acc=jnp.mean(jnp.argmax(logits, axis=-1) == labels["target"]),
            loss=loss,
        )
        loss = self.loss_scale.scale(loss)
        return loss, logs, model

    def train_step(self: M, inputs, labels) -> eg.TrainStepOutput[M]:
        model: M = self

        params = model.parameters()
        # train
        grads, (logs, model) = jax.grad(Model.loss_fn, has_aux=True)(
            params,
            model,
            inputs,
            labels,
        )
        grads = self.loss_scale.unscale(grads)

        assert model.optimizer is not None

        params = model.optimizer.update(grads, params)
        model = model.merge(params)

        return logs, model


def main(...):

    dataset = load_dataset("mnist")
    # dataset.set_format("np")
    def tfms(batch):
        batch["image"] = [jnp.array(img) for img in batch["image"]]
        return batch

    dataset.set_transform(tfms, columns=["image"])
    X_train = jnp.stack(dataset["train"]["image"])
    y_train = jnp.stack(dataset["train"]["label"])
    X_test = jnp.stack(dataset["test"]["image"])
    y_test = jnp.stack(dataset["test"]["label"])

    model = Model(
        features_out=10,
        optimizer=optax.adam(1e-3),
        eager=eager,
    )

    history = model.fit(...  )
    jax.profiler.save_device_memory_profile('memory-half.prof')
...
```
I tried to turn off memory pre-allocation and compare, in nvidia-smi but they turn out rougly to be the same. So I tried with the pprof profiler.

Memory profile train with half precision
<img width="718" alt="Screen Shot 2022-03-18 at 21 56 34" src="https://user-images.githubusercontent.com/12573521/159026823-317195b8-2ada-44f9-bdbe-24eef293e1f2.png">

Memory profile train with full precision
<img width="1092" alt="Screen Shot 2022-03-18 at 21 57 49" src="https://user-images.githubusercontent.com/12573521/159027085-244f97f6-2b52-400f-883f-fd919caa1a49.png">

**Comment by lkhphuc** [comment]
It seems the example does indeed use half precision for computation, but at some step it create a spike of memory just equal to the the full precision model, thus cannot fit larger model using half precision yet.
